src/NumericalAlgorithms/LinearSolver/Gmres.hpp

```diff
@@ -0,0 +1,209 @@
+      : convergence_criteria_(std::move(convergence_criteria)),  // NOLINT
+        verbosity_(std::move(verbosity)),                        // NOLINT
+        restart_(restart > 0 ? restart : convergence_criteria_.max_iterations) {
+    initialize();
+  }
+
+  Gmres() = default;
+  Gmres(const Gmres& /*rhs*/) = default;
+  Gmres& operator=(const Gmres& /*rhs*/) = default;
+  Gmres(Gmres&& /*rhs*/) noexcept = default;
+  Gmres& operator=(Gmres&& /*rhs*/) noexcept = default;
+  ~Gmres() = default;
+
+  void initialize() noexcept {
+    orthogonalization_history_.reserve(restart_ + 1);
+    basis_history_.resize(restart_);
+    preconditioned_basis_history_.resize(restart_);
+  }
+
+  const Convergence::Criteria& convergence_criteria() const noexcept {
+    return convergence_criteria_;
+  }
+
+  Convergence::Reason convergence_reason() const noexcept {
+    ASSERT(convergence_reason_,
+           "Tried to retrieve the convergence reason, but has not performed a "
+           "solve yet.");
+    return *convergence_reason_;
+  }
+
+  size_t num_iterations() const noexcept { return iteration_; }
+
+  void pup(PUP::er& p) noexcept {  // NOLINT
+    p | convergence_criteria_;
+    p | verbosity_;
+    p | restart_;
+    if (p.isUnpacking()) {
+      initialize();
+    }
+  }
+
+  template <typename LinearOperator, typename SourceType,
+            typename Preconditioner = IdentityPreconditioner<VarsType>>
+  VarsType operator()(LinearOperator&& linear_operator,
+                      const SourceType& source, const VarsType& initial_guess,
+                      const Preconditioner& preconditioner =
+                          IdentityPreconditioner<VarsType>{}) const noexcept {
+    constexpr bool use_preconditioner =
+        not cpp17::is_same_v<Preconditioner, IdentityPreconditioner<VarsType>>;
+
+    auto result = initial_guess;
+    convergence_reason_ = boost::none;
+    iteration_ = 0;
+    // Can allocated memory for the operand_ be kept around for successive
+    // solves?
+    VarsType operand_{};
+    while (not convergence_reason_) {
+      operand_ = source - linear_operator(result);
+      const double initial_residual_magnitude =
+          sqrt(inner_product(operand_, operand_));
+      convergence_reason_ = Convergence::criteria_match(
+          convergence_criteria_, iteration_, initial_residual_magnitude,
+          initial_residual_magnitude);
+      if (convergence_reason_) {
+        break;
+      }
+      operand_ /= initial_residual_magnitude;
+      basis_history_[0] = operand_;
+
+      std::pair<DenseVector<double>, double> minres_and_magnitude{};
+      for (size_t k = 0; k < restart_; k++) {
+        if (use_preconditioner) {
+          preconditioned_basis_history_[k] = preconditioner(operand_);
+        }
+        operand_ = linear_operator(
+            use_preconditioner ? preconditioned_basis_history_[k] : operand_);
+        // Resize matrix
+        orthogonalization_history_.resize(k + 2, k + 1);
```

Comment:
why not just allocate the matrix of the correct size to begin with and use `k+2` and `k+1` where you currently use `rows` and `columns`?

Reply:
If I understand Blaze correctly, that's what the `reserve` in the `initialize` function does. `resize` then sets the size being returned by `rows()` and `columns()`.

Reply:
Ah, good. I've overlooked the reservation in the initialization. No objections, then. :+1: